### PR TITLE
Build TrafficRoutes

### PR DESCRIPTION
**Pull request by ksubrmnn**
**Description**:
This PR creates new APIs/structs to build traffic routes in an extensible way as a first step to v1alpha3 traffic split. Only the API work has been done here, the routes are not yet consumed in any place in the code. 

This is a draft PR. There are still a few things to be resolved

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

File: pkg/catalog/routes.go [hunk 1]
```diff
@@ -333,12 +335,31 @@ func getTrafficPoliciesForService(mc *MeshCatalog, routePolicies map[trafficpoli
...
 	}
 
 	log.Debug().Msgf("Traffic policies for service %s: %+v", meshService, matchedTrafficTargets)
-	return matchedTrafficTargets, nil
+	return matchedTrafficTargets, nil*/
+
+	return srcDstTrafficTargetMap, nil
+}
+
+func (mc *MeshCatalog) getTrafficRoutesForService(svc service.MeshService, trafficTargets map[string]trafficpolicy.TrafficTarget) []trafficpolicy.TrafficRoutes {
```
Comment (ksubrmnn) [inline pkg/catalog/routes.go:348]: The arg for this function is what is returned by getTrafficPoliciesForService(). I will refactor a bit here after an initial review

File: pkg/catalog/routes.go [hunk 2]
```diff
@@ -397,3 +418,75 @@ func listTrafficTargetPermutations(name string, srcServiceList []service.MeshSer
...
+		routeWeightedSvcs = append(routeWeightedSvcs, rWS)
+	}
+
+	return routeWeightedSvcs
+}
+
+func getRouteKey(route trafficpolicy.HTTPRoute, routeMap map[*trafficpolicy.HTTPRoute][]service.WeightedService) *trafficpolicy.HTTPRoute {
+	var matchingKey *trafficpolicy.HTTPRoute
+
+	for key := range routeMap {
```
Comment (eduser25) [inline pkg/catalog/routes.go:484]: The map serves no purpose here I believe. 

You just do a linear search on the map to return the key, to later use the key as a map lookup index.
I have not looked into it deeply, but I'd probably suggest to find a way to index `HTTPRoute` by some usable key.

File: pkg/catalog/routes.go [hunk 3]
```diff
@@ -397,3 +418,75 @@ func listTrafficTargetPermutations(name string, srcServiceList []service.MeshSer
...
+
+	// Retrieve the weighted clusters from traffic split
+	// Assumes no nested splits for now
+	// Assumes service would only be root once (not sure if this is validated)
+	servicesList := mc.meshSpec.ListTrafficSplitServices()
+	for _, activeService := range servicesList {
+		if activeService.RootService == svc.Name {
+			weightedSvcs = append(weightedSvcs, activeService)
+		}
+	}
```
Comment (eduser25) [inline pkg/catalog/routes.go:433]: Correct me if wrong, this patch does not address https://github.com/openservicemesh/osm/issues/1798 right? (just making sure here)

File: pkg/catalog/routes.go [hunk 4]
```diff
@@ -397,3 +418,75 @@ func listTrafficTargetPermutations(name string, srcServiceList []service.MeshSer
...
+func (mc *MeshCatalog) getRouteWeightedServices(src service.MeshService, dest service.MeshService, trafficTargets map[string]trafficpolicy.TrafficTarget) []trafficpolicy.RouteWeightedServices {
+	routeWeightedSvcs := []trafficpolicy.RouteWeightedServices{}
+
+	weightedBackends := mc.getWeightedBackendsForService(dest)
+
+	routeServicesMap := make(map[*trafficpolicy.HTTPRoute][]service.WeightedService)
+
+	for _, backend := range weightedBackends {
+		hash := hashSrcDstService(src, backend.Service)
+		routes := trafficTargets[hash].HTTPRoutes
```
Comment (eduser25) [inline pkg/catalog/routes.go:454]: shouldn't `trafficTargets[hash]` be safeguarded and check it exists before accessing an element into it?
even if it were to be guaranteed by some other precondition, best practices to err check

File: pkg/tests/fixtures.go [hunk 5]
```diff
@@ -57,8 +64,11 @@ const (
 	// WildcardWithHeadersMatchName is the name of the match object.
 	WildcardWithHeadersMatchName = "allow-everything-on-header"
 
-	// Weight is the percentage of the traffic to be sent this way in a traffic split scenario.
-	Weight = 100
+	// Weight1 is the weight of the traffic to be sent this way in a traffic split scenario.
+	Weight1 = 100
+
+	// Weight2is the weight of the traffic to be sent this way in a traffic split scenario.
```
Comment (snehachhabria) [inline pkg/tests/fixtures.go:70]:
```suggestion
	// Weight2 is the weight of the traffic to be sent this way in a traffic split scenario.
```

File: pkg/tests/fixtures.go [hunk 6]
```diff
@@ -261,6 +264,35 @@ var (
 		},
 	}
 
+	// BookstoreV2TrafficTarget is a traffic target SMI object.
+	BookstoreV2TrafficTarget = target.TrafficTarget{
```
Comment (snehachhabria) [inline pkg/tests/fixtures.go:268]: Should we add another fixture and test where bookstorev1 and bookstorev2 have the same service account ?

File: pkg/catalog/routes.go [hunk 7]
```diff
@@ -333,12 +335,31 @@ func getTrafficPoliciesForService(mc *MeshCatalog, routePolicies map[trafficpoli
...
+		trafficRoute := trafficpolicy.TrafficRoutes{
+			//Name:                  target.Name,
+			Source:                target.Source,
+			Destination:           target.Destination,
+			RouteWeightedServices: mc.getRouteWeightedServices(target.Source, target.Destination, trafficTargets),
+		}
+		trafficRoutes = append(trafficRoutes, trafficRoute)
+	}
+
+	//TODO: Add roots of traffic splits (will need some slight testing changes)
```
Comment (snehachhabria) [inline pkg/catalog/routes.go:361]: Discussed offline to get all unique apex services and then build the traffic routes for these apex services

File: pkg/catalog/routes.go [hunk 8]
```diff
@@ -333,12 +335,31 @@ func getTrafficPoliciesForService(mc *MeshCatalog, routePolicies map[trafficpoli
 		}
 	}
 
-	for _, trafficTarget := range srcDstTrafficTargetMap {
+	/*for _, trafficTarget := range srcDstTrafficTargetMap {
```
[the remark this answers is not in the document]
Comment (ksubrmnn) [reply pkg/catalog/routes.go:338]: The function takes MeshService so the hash is already namespaced

See the [test](https://github.com/openservicemesh/osm/blob/main/pkg/catalog/routes_test.go#L315) for an example

**Pull request closed without merging**